swank/abcl.lisp

```diff
@@ -541,12 +541,19 @@
 (defmethod source-location ((symbol symbol))
   (when (pathnamep (ext:source-pathname symbol))
     (let ((pos (ext:source-file-position symbol)))
-      `(:location
-        (:file ,(namestring (ext:source-pathname symbol)))
-        ,(if pos
-             (list :position (1+ pos))
-             (list :function-name (string symbol)))
-        (:align t)))))
+      (if (ext:pathname-jar-p (ext:source-pathname symbol))
+          (destructuring-bind (jar entry) (abcl-asdf::split-string (subseq (namestring (ext:source-pathname symbol)) 9) (char "!" 0))
+            `(:location (:zip ,jar ,(subseq entry 1))
+                        ,(if nil ;pos -- positions seem to be wrong. Just use function name
+                            (list :position (1+ pos))
+                            (list :function-name (string symbol)))
+                        (:align t)))
+          `(:location
+            (:file ,(namestring (ext:source-pathname symbol)))
+            ,(if nil;pos
+                 (list :position (1+ pos))
+                 (list :function-name (string symbol)))
+            (:align t))))))
```

Comment:
There's a lot of duplication between these two location blocks. Seems like it should be easily refactored.